### PR TITLE
Remove golint based verifications

### PR DESCRIPTION
**Pull request by dims**
After a couple years of seeing changes spawned from this issue, I don't think this is a good use of contributors' or reviewers' time. golint is very opinionated about:

exported type and function names, leading to PRs attempting to change exported things people use downstream
missing documentation on interface implementation functions, which leads to not-very-useful copy/paste of interface godoc
the govet and staticcheck verify jobs are much more useful since they catch correctness issues

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
Please see notes in https://github.com/kubernetes/kubernetes/issues/68026#issuecomment-760193686

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
also drop golint from `hack/tools/Makefile` and `hack/tools/tools.go`?

File: hack/verify-golint.sh
Comment (BenTheElder) [inline hack/verify-golint.sh:3]: an oldy 🙃

**Comment by BenTheElder** [comment]
There are some other golint references outside of vendor:
https://cs.k8s.io/?q=golint&i=nope&files=%5E%5B%5Ev%5D&repos=kubernetes/kubernetes

In particular this one seems worth looking at https://github.com/kubernetes/kubernetes/blob/9da11e294fae8e26330d5c0ed5dff61cda178d77/hack/verify-staticcheck.sh#L39

**Review by spiffxp — COMMENTED** [review]
/hold
I'm guessing this is going to cause ~35 open PRs to have `needs-rebase` applied (ref: https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+is%3Aopen+%22golint%22+)

I agree our review bandwidth is better spent elsewhere, but I think it'd be rude to remove this without some kind of notice.  AFAICT we've discussed this for ~12 hours.

An alternative first step could be to exclude golint from the list of verify scripts that are run as part of `make verify`, thus making it non-blocking (e.g. add to https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/verify.sh#L33)

You are welcome to remove the /hold if you would rather proceed as-is.

**Comment by dominikh** [comment]
> @BenTheElder if i remove that line, we have a bunch of failures (full output is http://paste.openstack.org/raw/801644/ )

`ST` is Staticcheck's own set of style checks. It overlaps with golint, but is neither a subset nor a superset. If you want to stop checking for style issues altogether, you'd best keep the ST* checks disabled, as you currently do. Alternatively, you could look through all available style checks (https://staticcheck.io/docs/checks#ST1000) and only enable those that you agree with. Unlike golint, Staticcheck lets you pick exactly the checks that you want.

**Comment by fejta** [comment]
Personally I'd rather see the lint checks enabled everywhere -- at which point the code will be easier to maintain and not waste reviewers time.

The current state of having part of the code lint-free and other parts with lint errors is the worst of both worlds.

**Comment by dims** [comment]
there's a thread on k-dev as well:
https://groups.google.com/g/kubernetes-dev/c/q13Y2Xo2sLI/m/C9tfLLNlDgAJ

**Comment by justaugustus** [comment]
Agreed with @fejta on seeing linting everywhere.
We have a [pretty strict config](https://github.com/kubernetes/release/blob/master/.golangci.yml) in k/release, mostly because I expected to make a ton of idiomatic mistakes while learning.

Would love to see the replacement land (https://github.com/kubernetes/kubernetes/pull/93656) before we rip this out.

**Comment by dims** [comment]
@justaugustus #93656 is not going to land unfortunately, there are too many things that need to be fixed before that can land. (same stuff  listed as broken in the failures file is just the tip of the iceberg).

**Comment by camilamacedo86** [comment]
Hi folks, 

If the problem is to do reviews to ensure that code is passing in the lint checks what about to use https://github.com/golangci/golangci-lint with the GA such as [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/.github/workflows/ci.yml#L11-L23). See that the PR will be checked automatically and if issues are found then comments will be raised on it to help the author know what is missing and fix it. See:

screen-shot-2020-08-21-at-17 19 57


Regarding the point 
> One question is do we have to block merging PRs which fix coding-style?

IMO yes, in order to : 

- Ensure code readability
- Maintainability
- Help the code review process
- Avoid (syntax) issues
- Ensure styles and standards
- Following the good practices

**Comment by BenTheElder** [comment]
I think for `lint` specifically it's important to reference their documentation:

https://github.com/golang/lint#purpose

> The suggestions made by golint are exactly that: suggestions. Golint is not perfect, and has both false positives and false negatives. Do not treat its output as a gold standard. We will not be adding pragmas or other knobs to suppress specific warnings, so do not expect or require code to be completely "lint-free". In short, this tool is not, and will never be, trustworthy enough for its suggestions to be enforced automatically, for example as part of a build process. Golint makes suggestions for many of the mechanically checkable items listed in Effective Go and the CodeReviewComments wiki page.

TLDR: the `lint` tool specifically  is not meant to be mechanically enforced, it's supposed to be used to provide suggestions while developing. 

We have other tools like staticcheck, gofmt that are actually meant to be enforced that we enforce already.

---


> If the problem is to do reviews to ensure that code is passing in the lint checks what about to use https://github.com/golangci/golangci-lint with the GA such as here. See that the PR will be checked automatically and if issues are found then comments will be raised on it to help the author know what is missing and fix it. See:

GitHub Actions does not work well with other aspects of how we do CI + merge in Kubernetes, namely that we ensure that tests ran against the latest of (target branch + PR branch) prior to merge.

[Our CI](https://github.com/kubernetes/test-infra/tree/master/prow) actually has [the ability to comment golint suggestions](https://github.com/kubernetes/test-infra/tree/master/prow/plugins/golint) as well, but that feature isn't widely used currently. We could consider enabling it on PRs without requiring the `/lint` comment trigger (perhaps using a different account to avoid rate-limit issues).

> > One question is do we have to block merging PRs which fix coding-style?
> IMO yes, in order to :

Er, I think you've read this backwards, they're asking if we need to *not* merge PRs which fix lints, not block PRs that _need_ to fix lints?

We shouldn't need to.

> Btw, why the CI should allow for example merge a PR with unused code or with fmt issues if it reduces the maintainability of the project?

We don't. We run a lot of other lint / static analysis. PRs that are not formatted will not merge. `make verify` covers a lot including `gofmt`, `staticcheck`, etc. -- You can get some idea with `ls hack/verify-*` or check:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98063/pull-kubernetes-verify/1352698838279786496
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98063/pull-kubernetes-verify-govet-levee/1352698837461897216
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98063/pull-kubernetes-typecheck/1352698836610453504

(test result statuses from this PR, with more details about what ran).

**Comment by fejta** [comment]
LGTM
/uncc

**Comment by dims** [comment]
folks have been notified on the issue, over email to -dev and sufficient time has passed.

@spiffxp i'd like to proceed with this. removing the hold

/hold cancel

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
